### PR TITLE
Add few unit test cases in ASSERT_STATUS_CHECKED build

### PR DESCRIPTION
**Pull request by akankshamahajan15**
Summary: Fix few test cases and add them in ASSERT_STATUS_CHECKED build.

Test Plan: 1.  ASSERT_STATUS_CHECKED=1 make -j48 check,
                 2. travis build for ASSERT_STATUS_CHECKED,
                 3. Without ASSERT_STATUS_CHECKED:  make check -j64, CircleCI build and travis build

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: db/compaction/compaction_job.cc [hunk 1]
```diff
@@ -1024,6 +1025,8 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
       RecordDroppedKeys(range_del_out_stats,
                         &sub_compact->compaction_job_stats);
     }
+    // Status is checked in while loop.
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:1028]: Since I agree that the status is checked in the while loop, I am wondering why these lines are necessary?

File: db/compaction/compaction_job.cc [hunk 2]
```diff
@@ -1640,6 +1643,7 @@ void CompactionJob::CleanupCompaction() {
         TableCache::Evict(table_cache_.get(), out.meta.fd.GetNumber());
       }
     }
+    sub_compact.io_status.PermitUncheckedError();
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:1646]: Is there a reason not to do this in the SubCompactionState destructor instead?  If not, please add a comment why this is okay here (because the sub_compact is about to be deleted anyway...)

File: db/db_impl/db_impl_compaction_flush.cc [hunk 3]
```diff
@@ -1259,6 +1261,7 @@ void DBImpl::NotifyOnCompactionBegin(ColumnFamilyData* cfd, Compaction* c,
     BuildCompactionJobInfo(cfd, c, st, job_stats, job_id, current, &info);
     for (auto listener : immutable_db_options_.listeners) {
       listener->OnCompactionBegin(this, info);
+      info.status.PermitUncheckedError();
```
Comment (mrambacher) [inline db/db_impl/db_impl_compaction_flush.cc:1264]: This should probably be outside of the for loop

File: db/merge_helper.cc [hunk 4]
```diff
@@ -140,15 +140,17 @@ Status MergeHelper::MergeUntil(InternalIterator* iter,
   ParsedInternalKey orig_ikey;
   bool succ = ParseInternalKey(original_key, &orig_ikey);
   assert(succ);
+  Status s;
```
Comment (mrambacher) [inline db/merge_helper.cc:143]: Are these changes necessary?

File: db/column_family.cc [hunk 5]
```diff
@@ -1102,10 +1102,10 @@ Status ColumnFamilyData::RangesOverlapWithMemtables(
       super_version->mem->NewRangeTombstoneIterator(read_opts, read_seq);
   range_del_agg.AddTombstones(
       std::unique_ptr<FragmentedRangeTombstoneIterator>(active_range_del_iter));
-  super_version->imm->AddRangeTombstoneIterators(read_opts, nullptr /* arena */,
-                                                 &range_del_agg);
-
   Status status;
+  status = super_version->imm->AddRangeTombstoneIterators(
```
Comment (mrambacher) [inline db/column_family.cc:1106]: Since AddRangeTombstoneIterators always returns OK, maybe just change it to return a void instead?

**Review by mrambacher — COMMENTED** [review]
I cannot tell from your test plan, but you also need to run the tests without ASSERT_STATUS_CHECKED set, to verify the tests pass that way as well.

File: db/column_family.cc (hunk 5 above)
Comment (akankshamahajan15) [reply db/column_family.cc:1106]: Yes, I wasn't sure if I should  change the API's (return type) for this change.

File: db/merge_helper.cc (hunk 4 above)
Comment (akankshamahajan15) [reply db/merge_helper.cc:143]: It gives error because Status s is defined but its status is not checked in the flow where direct status is returned  instead of s. So to make it consistent I assigned the error messages to s and returned s.

**Comment by akankshamahajan15** [comment]
> I cannot tell from your test plan, but you also need to run the tests without ASSERT_STATUS_CHECKED set, to verify the tests pass that way as well.

Thank you for pointing that out. I updated the test plan.

File: db/compaction/compaction_job.cc (hunk 1 above)
Comment (akankshamahajan15) [reply db/compaction/compaction_job.cc]: I checked again, it's not needed so removed it.

**Comment by akankshamahajan15** [comment]
build-windows-4 failure not related to this change.

File: db/column_family.cc (hunk 5 above)
Comment (mrambacher) [reply db/column_family.cc:1106]: Can you add a comment or PR that says the method always returns OK so it can be addressed sometime later?

**Review by mrambacher — APPROVED** [review]
If it passes the tests, it LGTM.  The other comments I had are fairly minor.
